crates/sui-json-rpc/src/error.rs

```diff
@@ -161,12 +162,28 @@ impl From<Error> for RpcError {
                 QuorumDriverError::NonRecoverableTransactionError { errors } => {
                     let new_errors: Vec<String> = errors
                         .into_iter()
-                        .filter(|(err, _, _)| !err.is_retryable().0) // consider retryable errors as transient errors
-                        .map(|(err, _, _)| {
-                            match err {
-                                // TODO(wlmyng): update SuiError display trait to render UserInputError with display
-                                SuiError::UserInputError { error } => error.to_string(),
-                                _ => err.to_string(),
+                        // sort by total stake, descending, so users see the most prominent one first
+                        .sorted_by(|(_, a, _), (_, b, _)| b.cmp(a))
+                        .filter_map(|(err, _, _)| {
+                            match &err {
+                                // Special handling of UserInputError:
+                                // ObjectNotFound and DependentPackageNotFound are considered
+                                // retryable errors but they have different treatment
+                                // in AuthorityAggregator.
+                                // The optimal fix would be to examine if the total stake
+                                // of ObjectNotFound/DependentPackageNotFound exceeds the
+                                // quorum threshold, but it takes a Committee here.
+                                // So, we take an easier route and consider them non-retryable
+                                // at all. Combining this with the sorting above, clients will
+                                // see the dominant error first.
+                                SuiError::UserInputError { error } => Some(error.to_string()),
+                                _ => {
+                                    if err.is_retryable().0 {
```

Comment:
I might be missing something but isn't this the same logic as !error.is_retryable()? Should there be an accompanying change in the is_retryable method to set ObjectNotFound as not retryable

Reply:
No we will not change `is_retryable`. They are still retryable from AuthorityAggregator's perspective